Strategy Design Pattern

katonaaron-msg opened the pull request:
Apply Strategy Design Pattern for Order Creation
- Implement two LocationSelectionAlgorithms and update OrderService to use them.
- Create a bean for the algorithm with LocationSelectionConfiguration
- Read the LocationSelectionAlgorithm implementation's name from the properties file
- Move shippedFrom location from Order to OrderDetail. Migrate DB.

katonaaron-msg committed:
Apply Strategy Design Pattern for Order Creation

- Implement two LocationSelectionAlgorithms and update OrderService to use them.
- Create a bean for the algorithm with LocationSelectionConfiguration
- Read the LocationSelectionAlgorithm implementation's name from the properties file
- Move shippedFrom location from Order to OrderDetail. Migrate DB.

cristinaAnca commented on src/main/java/ro/msg/learning/shop/exception/LocationOutOfStockException.java:
```diff
@@ -0,0 +1,21 @@
+
+@Getter
+public class LocationOutOfStockException extends RuntimeException {
+    private final UUID locationId;
```
Do you really need these variables? Do you use them elsewhere than the constructor in order to set the error message?

katonaaron-msg replied:
Removed them

cristinaAnca commented on src/main/java/ro/msg/learning/shop/model/Location.java:
```diff
@@ -27,4 +29,21 @@ public void setStocks(Set<Stock> stocks) {
         stocks.forEach(stock -> stock.setLocation(this));
     }
+
+    public void removeProductFromStock(@NonNull Product product, @NonNull Integer quantity) {
```
Please keep the model entities as POJO (no implementation inside the entities). This method should be part of a service class.

cristinaAnca commented on src/main/java/ro/msg/learning/shop/service/locationselection/MostAbundantLocationSelectionAlgorithm.java:
```diff
@@ -0,0 +1,34 @@
+import ro.msg.learning.shop.model.OrderDetail;
+import ro.msg.learning.shop.model.Stock;
+
+import java.util.*;
```
Replace * imports with explicit imports. We don't want to import everything from java.util. Please replace also all the * imports from the entire project with explicit import

cristinaAnca commented on src/main/resources/application.properties:
```diff
@@ -4,4 +4,6 @@ spring.datasource.password=postgres
-spring.jpa.properties.hibernate.format_sql=true
\ No newline at end of file
+spring.jpa.properties.hibernate.format_sql=true
+#ro.msg.learning.shop.location-selection-algorithm=ro.msg.learning.shop.service.locationselection.SingleLocationSelectionAlgorithm
```
Remove this comment

katonaaron-msg committed:
Fix issues

- remove wildcard import, comment in properties file
- moved logic from entity to service
- simplified the LocationOutOfStockException class
